internal/storage/v2/memory/memory.go

```diff
@@ -116,15 +116,30 @@ func (st *Store) GetServices(ctx context.Context) ([]string, error) {
 
 func (st *Store) FindTraces(ctx context.Context, query tracestore.TraceQueryParams) iter.Seq2[[]ptrace.Traces, error] {
 	m := st.getTenant(tenancy.GetTenant(ctx))
-	if query.SearchDepth <= 0 || query.SearchDepth > st.defaultConfig.MaxTraces {
-		return func(yield func([]ptrace.Traces, error) bool) {
-			yield(nil, errInvalidSearchDepth)
+	return func(yield func([]ptrace.Traces, error) bool) {
+		traceAndIds, err := m.findTraceAndIds(query)
+		if err != nil {
+			yield(nil, err)
+			return
+		}
+		for i := range traceAndIds {
+			if !yield([]ptrace.Traces{traceAndIds[i].trace}, nil) {
+				return
+			}
 		}
 	}
-	traces := m.findTraces(query)
-	return func(yield func([]ptrace.Traces, error) bool) {
-		for _, trace := range traces {
-			if !yield([]ptrace.Traces{trace}, nil) {
+}
+
+func (st *Store) FindTraceIDs(ctx context.Context, query tracestore.TraceQueryParams) iter.Seq2[[]tracestore.FoundTraceID, error] {
+	m := st.getTenant(tenancy.GetTenant(ctx))
+	return func(yield func([]tracestore.FoundTraceID, error) bool) {
+		traceAndIds, err := m.findTraceAndIds(query)
+		if err != nil {
+			yield(nil, err)
+			return
+		}
+		for i := range traceAndIds {
+			if !yield([]tracestore.FoundTraceID{{TraceID: traceAndIds[i].id}}, nil) {
```

Comment:
this is overkill, what you had before was better, I was not suggesting to change that part. Collect an array of IDs and return it in a single call to yield.